### PR TITLE
Download BoM files

### PR DESCRIPTION
**Pull request by AbdulrhmnGhanem**
1. Use https://github.com/kasbah/push-on-hold-off as a test fixture.
2. Migrate all the jsx files I touched to tsx
3. Add Plausible types
4. Download bom csv file optimizied for each retailer
5. Extract some inline styles into scss files

(unit tests for the new download feature and digikey's direct link)

##
## templates
1. Digikey remains a link
7. [Mouser.xls](https://github.com/user-attachments/files/17264725/BomTemplate.xls)
8. [RS.xls](https://github.com/user-attachments/files/17264730/RS-quote-template-xa.xls)
9. [Farnell.xlsx](https://github.com/user-attachments/files/17264732/Sample_Template_en_GB.xlsx)
10. Newark has the same file format as Farnell

### CODE REVIEW

**Comment by kasbah** [comment]
Stick to `.csv` if you can. We don't want to load an XLSX library client side. Here is [LCSC_BomTemplate.csv](https://github.com/user-attachments/files/17266296/LCSC_BomTemplate.csv).

**Comment by AbdulrhmnGhanem** [comment]
Sure, all of them accept csv but the templates are xlsx

**Comment by kasbah** [comment]
One thing to watch out for is the Unicode encoding for the downloaded `.csv`. 

- Javascript might default to UTF-16, I can't remember. 
- Excel defaults to UTF-16 I'm pretty sure. 
- Does the browser do different things on different platforms? 

Should probably test downloading and uploading to vendor some BOMs with unicode characters on different platforms.

**Comment by AbdulrhmnGhanem** [comment]
I will do it as CSV first and test. If I find it too hard to stratify every vendor, I will switch to XLSX.

**Comment by AbdulrhmnGhanem** [comment]
The RS template is out of date, I downloaded the real BoM for a project and made the implementation according to it [cart.csv](https://github.com/user-attachments/files/17285788/cart.csv)

**Comment by AbdulrhmnGhanem** [comment]
Is there a board that has the six retailers in the BoM? I want to use it as a test fixture.

**Comment by kasbah** [comment]
Maybe  not, I can only think of https://kitspace.org/boards/github.com/kasbah/push-on-hold-off/ which has 5. If you send a PR for LCSC I'll merge it though :)

**Comment by AbdulrhmnGhanem** [comment]
>  If you send a PR for LCSC I'll merge it though :)

https://github.com/kasbah/push-on-hold-off/pull/7

**Comment by AbdulrhmnGhanem** [comment]
commit 7708f29

|before|after|
|---|---|
|image|image|

File: frontend/src/components/Board/BuyParts/Bom.tsx [hunk 1]
```diff
@@ -44,13 +43,13 @@ const ExpandBom = ({ diff, collapsed, setCollapsed }) => {
     ) : null
 
   return (
-    <div style={{ paddingLeft: 1, paddingRight: 1 }}>
+    <div className={`${styles.expandWrapper}`}>
```
Comment (kasbah) [inline frontend/src/components/Board/BuyParts/Bom.tsx:46]:
```suggestion
    <div className={styles.expandWrapper}>
```

File: frontend/src/components/Board/BuyParts/index.module.scss [hunk 2]
```diff
@@ -1,8 +1,10 @@
 @import '../../breakpoint.scss';
 
 .notSelectable {
+  -webkit-user-select: none;
   user-select: none;
   cursor: default;
+  margin-left: 5px;
```
Comment (kasbah) [inline frontend/src/components/Board/BuyParts/index.module.scss:7]: Is the -webkit prefix one needed?

Why are we mixing margin concerns with notSelectable?

**Review by kasbah — CHANGES_REQUESTED** [review]
The buttons are always green now when before they would be red when not all parts are in the BOM for that store.

File: frontend/src/components/Board/BuyParts/index.tsx [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+function downloadBomCSV({
+  name,
+  lines,
+  multiplier,
+  buyAddPercent,
+  retailer,
+}: Order) {
+  const csvContent = `${csvBom(lines, multiplier, buyAddPercent, retailer)
+    .map((e: Array<string>) => e.join(','))
+    .join('\n')}\n`
```
Comment (kasbah) [inline frontend/src/components/Board/BuyParts/index.tsx:397]: - Strings with commas in them need to be quoted
- Strings with quotes in them need to be escaped

File: frontend/src/components/Board/BuyParts/index.tsx [hunk 4]
```diff
@@ -408,6 +408,20 @@ function redirectToStore(retailer: Retailer) {
...
+/**
+  Escape a CSV value according to RFC 4180.
+  https://www.ietf.org/rfc/rfc4180.txt
+ */
+function escapeCsvValue(value: string | number | undefined) {
+  if (
+    typeof value === 'string' &&
+    (value.includes(',') || value.includes('"') || value.includes('\n'))
+  ) {
+    return `"${value.replace(/"/g, '""')}"`
```
Comment (kasbah) [inline frontend/src/components/Board/BuyParts/index.tsx:420]: Is that right, shouldn't it be `\"`?

File: frontend/src/components/Board/BuyParts/index.tsx [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+
+function downloadBomCSV({
+  name,
+  lines,
+  multiplier,
+  buyAddPercent,
+  retailer,
+}: Order) {
+  const csvContent = `${csvBom(lines, multiplier, buyAddPercent, retailer)
+    .map((e: Array<CsVValue>) => e.map(escapeCsvValue).join(','))
```
Comment (kasbah) [inline frontend/src/components/Board/BuyParts/index.tsx:396]:
```suggestion
    .map((e: Array<CsvValue>) => e.map(escapeCsvValue).join(','))
```

File: frontend/src/components/Board/BuyParts/index.tsx [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+  totalLines: number
+  numberOfParts: number
+  icons: 'link' | 'download'
+}
+
+interface StoreIconProps {
+  retailer: string
+}
+
+type CsVValue = string | number | undefined
```
Comment (kasbah) [inline frontend/src/components/Board/BuyParts/index.tsx:459]:
```suggestion
type CsvValue = string | number | undefined
```

File: frontend/src/components/Board/BuyParts/index.tsx (hunk 4 above)
Comment (AbdulrhmnGhanem) [reply frontend/src/components/Board/BuyParts/index.tsx:420]: I implemented it according to the RFC linked in the function docstring

image

**Review by kasbah — APPROVED** [review]

**Pull request merged**